File: php-server/prompt/case_file_analyze.php
<?php
declare(strict_types=1);

require_once __DIR__ . '/../inc/db.php';
require_once __DIR__ . '/../inc/auth.php';

ini_set('display_errors', '0');
ini_set('max_execution_time', '180');
set_time_limit(180);
header('Content-Type: application/json; charset=utf-8');

function fa_json_exit(array $payload): void
{
    if (ob_get_level() > 0) ob_clean();
    echo json_encode($payload, JSON_UNESCAPED_UNICODE);
    exit;
}

ob_start();
register_shutdown_function(static function (): void {
    $error = error_get_last();
    if ($error && in_array((int)$error['type'], [E_ERROR, E_PARSE, E_CORE_ERROR, E_COMPILE_ERROR], true)) {
        fa_json_exit(['error' => 'AI 처리 중 서버 오류가 발생했습니다.']);
    }
});

require_login();
$member   = current_member();
$memberPk = (int)($member['id'] ?? 0);

if ($_SERVER['REQUEST_METHOD'] !== 'POST') {
    fa_json_exit(['error' => '잘못된 요청입니다.']);
}

$caseType = trim((string)($_POST['case_type'] ?? ''));
$validTypes = ['problem_solve', 'process_work', 'consulting_qa', 'review_experience'];
if (!in_array($caseType, $validTypes, true)) {
    fa_json_exit(['error' => '올바른 사례 유형이 필요합니다.']);
}

if (empty($_FILES['ref_file']) || ($_FILES['ref_file']['error'] ?? UPLOAD_ERR_NO_FILE) !== UPLOAD_ERR_OK) {
    fa_json_exit(['error' => '파일이 첨부되지 않았습니다.']);
}

$file    = $_FILES['ref_file'];
$tmpPath = (string)($file['tmp_name'] ?? '');
$origName = (string)($file['name'] ?? '');
$fileSize = (int)($file['size'] ?? 0);

if ($fileSize > 5 * 1024 * 1024) {
    fa_json_exit(['error' => '파일 크기는 5MB 이하만 지원합니다.']);
}

$ext = strtolower(pathinfo($origName, PATHINFO_EXTENSION));
$allowedExts = ['txt', 'md', 'csv', 'log', 'json', 'html', 'htm', 'xml', 'rtf'];

$fileContent = '';

if (in_array($ext, $allowedExts, true)) {
    $fileContent = @file_get_contents($tmpPath);
    if ($fileContent === false) {
        fa_json_exit(['error' => '파일을 읽을 수 없습니다.']);
    }
    $fileContent = mb_convert_encoding($fileContent, 'UTF-8', 'auto');
    $fileContent = strip_tags($fileContent);
} elseif ($ext === 'pdf') {
    $projectRoot = realpath(__DIR__ . '/..');
    $venvPython = '';
    $candidates = [
        ($projectRoot !== false) ? rtrim($projectRoot, '/') . '/env/bin/python3' : '',
        '/usr/share/nginx/html/env/bin/python3',
    ];
    foreach ($candidates as $path) {
        if ($path !== '' && is_file($path)) {
            $venvPython = $path;
            break;
        }
    }
    if ($venvPython === '') $venvPython = 'python3';

    $pdfScript = 'import sys; '
        . 'from PyPDF2 import PdfReader; '
        . 'r = PdfReader(sys.argv[1]); '
        . 'print("\\n".join(p.extract_text() or "" for p in r.pages))';
    $cmd = escapeshellarg($venvPython) . ' -c ' . escapeshellarg($pdfScript) . ' ' . escapeshellarg($tmpPath) . ' 2>&1';
    $fileContent = trim((string)shell_exec($cmd));

    if ($fileContent === '' || mb_strlen(preg_replace('/\s+/', '', $fileContent)) < 30) {
        $ocrScript = __DIR__ . '/pdf_vision_ocr.py';
        if (is_file($ocrScript)) {
            $ocrCmd = escapeshellarg($venvPython) . ' ' . escapeshellarg($ocrScript) . ' ' . escapeshellarg($tmpPath) . ' 3 2>&1';
            $ocrOutput = trim((string)shell_exec($ocrCmd));
            if ($ocrOutput !== '') {
                $fileContent = $ocrOutput;
            }
        }
    }

    if ($fileContent === '') {
        fa_json_exit(['error' => 'PDF에서 텍스트를 추출할 수 없습니다. 텍스트 기반 PDF이거나 스캔 품질이 충분해야 합니다.']);
    }
} elseif (in_array($ext, ['doc', 'docx'], true)) {
    $projectRoot = realpath(__DIR__ . '/..');
    $venvPython = '';
    $candidates = [
        ($projectRoot !== false) ? rtrim($projectRoot, '/') . '/env/bin/python3' : '',
        '/usr/share/nginx/html/env/bin/python3',
    ];
    foreach ($candidates as $path) {
        if ($path !== '' && is_file($path)) {
            $venvPython = $path;
            break;
        }
    }
    if ($venvPython === '') $venvPython = 'python3';

    $docScript = 'import sys; '
        . 'from docx import Document; '
        . 'd = Document(sys.argv[1]); '
        . 'print("\\n".join(p.text for p in d.paragraphs))';
    $cmd = escapeshellarg($venvPython) . ' -c ' . escapeshellarg($docScript) . ' ' . escapeshellarg($tmpPath) . ' 2>&1';
    $fileContent = trim((string)shell_exec($cmd));
    if ($fileContent === '') {
        fa_json_exit(['error' => 'Word 파일 텍스트 추출에 실패했습니다.']);
    }
} elseif (in_array($ext, ['jpg', 'jpeg', 'png', 'gif', 'webp', 'bmp', 'tiff', 'tif'], true)) {
    $projectRoot = realpath(__DIR__ . '/..');
    $venvPython = '';
    $candidates = [
        ($projectRoot !== false) ? rtrim($projectRoot, '/') . '/env/bin/python3' : '',
        '/usr/share/nginx/html/env/bin/python3',
    ];
    foreach ($candidates as $path) {
        if ($path !== '' && is_file($path)) {
            $venvPython = $path;
            break;
        }
    }
    if ($venvPython === '') $venvPython = 'python3';

    $imgOcrScript = 'import os, sys, base64; '
        . 'from openai import OpenAI; '
        . 'from dotenv import load_dotenv; '
        . 'load_dotenv(os.path.join(os.path.dirname(os.path.abspath("' . addslashes(__DIR__) . '")), "api", ".env")); '
        . 'c = OpenAI(api_key=os.getenv("OPENAI_API_KEY","")); '
        . 'b = base64.b64encode(open(sys.argv[1],"rb").read()).decode(); '
        . 'r = c.chat.completions.create(model="gpt-4o-mini", messages=[{"role":"user","content":[{"type":"text","text":"이 이미지에 보이는 모든 텍스트를 빠짐없이 그대로 읽어서 출력해주세요. 구조를 유지하고 설명은 추가하지 마세요."},{"type":"image_url","image_url":{"url":"data:image/png;base64,"+b,"detail":"high"}}]}], temperature=0.1, max_tokens=4000); '
        . 'print(r.choices[0].message.content or "")';
    $cmd = escapeshellarg($venvPython) . ' -c ' . escapeshellarg($imgOcrScript) . ' ' . escapeshellarg($tmpPath) . ' 2>&1';
    $fileContent = trim((string)shell_exec($cmd));
    if ($fileContent === '') {
        fa_json_exit(['error' => '이미지에서 텍스트를 읽을 수 없습니다.']);
    }
} else {
    fa_json_exit(['error' => '지원하지 않는 파일 형식입니다. (txt, pdf, docx, csv, md, jpg, png 등을 사용해주세요)']);
}

$fileContent = trim($fileContent);
if (mb_strlen($fileContent) < 20) {
    fa_json_exit(['error' => '파일에서 충분한 텍스트를 추출할 수 없습니다.']);
}

$brandName = '';
$industry  = '';
if ($memberPk > 0) {
    try {
        $pdo  = db();
        $stmt = $pdo->prepare('SELECT brand_name, industry FROM caify_prompt WHERE member_pk = :member_pk LIMIT 1');
        $stmt->execute([':member_pk' => $memberPk]);
        $pRow = $stmt->fetch();
        if (is_array($pRow)) {
            $brandName = trim((string)($pRow['brand_name'] ?? ''));
            $industry  = trim((string)($pRow['industry'] ?? ''));
        }
    } catch (\Throwable $e) { }
}

$data = [
    'case_type'    => $caseType,
    'file_content' => $fileContent,
    'brand_name'   => $brandName,
    'industry'     => $industry,
];

$projectRoot = realpath(__DIR__ . '/..');
$pyScript    = __DIR__ . '/case_file_analyze.py';

$venvPython = '';
$candidates = [
    ($projectRoot !== false) ? rtrim($projectRoot, '/') . '/env/bin/python3' : '',
    '/usr/share/nginx/html/env/bin/python3',
];
foreach ($candidates as $path) {
    if ($path !== '' && is_file($path)) {
        $venvPython = $path;
        break;
    }
}
if ($venvPython === '') $venvPython = 'python3';

$cmd = escapeshellarg($venvPython) . ' ' . escapeshellarg($pyScript) . ' '
     . escapeshellarg(base64_encode(json_encode($data, JSON_UNESCAPED_UNICODE))) . ' 2>&1';

$output = trim((string)shell_exec($cmd));
if ($output === '') {
    fa_json_exit(['error' => 'AI 분석 스크립트 실행에 실패했습니다.']);
}

$response = json_decode($output, true);
if (!is_array($response)) {
    fa_json_exit(['error' => 'AI 응답 파싱 실패', 'raw' => mb_substr($output, 0, 500)]);
}

if (isset($response['response'])) {
    fa_json_exit(['success' => true, 'data' => $response['response']]);
} elseif (isset($response['error'])) {
    fa_json_exit(['error' => $response['error']]);
} else {
    fa_json_exit(['error' => 'AI 응답 처리 오류', 'raw' => mb_substr($output, 0, 500)]);
}

File: php-server/prompt/case_file_analyze.py
# -*- coding: utf-8 -*-
import sys
import json
import base64
import re

from case_short_ai import generate_response


def strip_markdown_codeblock(text: str) -> str:
    text = str(text or "").strip()
    m = re.match(r"^```(?:json)?\s*\n?(.*?)\n?\s*```$", text, re.DOTALL)
    return m.group(1).strip() if m else text


def build_analyze_prompt(data: dict) -> str:
    file_content = str(data.get("file_content", "")).strip()

    return f"""
당신은 문서 분석 전문가입니다.
사용자가 첨부한 문서를 읽고, 이 문서의 내용을 블로그 사례글로 만들기 위한 정보를 정리해주세요.

⚠️ 핵심 규칙:
- 오직 아래 문서에 적혀 있는 내용만 사용하세요.
- 문서에 없는 사실, 수치, 이름, 결과를 지어내지 마세요.
- 사례 유형(문제 해결, 작업 과정, 상담 등)을 미리 정하지 마세요. 문서 내용이 무엇이든 그대로 분석하세요.
- 문서가 평가서든, 보고서든, 메모든, 후기든 상관없이 있는 그대로의 내용을 정리하세요.

[첨부 문서 내용]
{file_content[:8000]}

요구사항:

1. summary: 이 문서가 어떤 내용인지 3~5문장으로 요약하세요. 문서에 있는 사실만 적으세요.

2. suggested_title: 이 문서 내용으로 블로그 사례글을 쓴다면 어울리는 제목을 제안하세요. (15~30자)

3. questions: 이 문서의 내용을 블로그 사례글로 만들기 위해 가장 적합한 질문 3개를 만들어주세요.
   - 질문은 문서에 실제로 답이 있는 내용을 기반으로 만드세요.
   - "문제가 뭐였나요?" 같은 일반적인 틀이 아니라, 이 문서의 내용에 딱 맞는 구체적인 질문이어야 합니다.
   - 예를 들어 문서가 인사 평가서면 "어떤 성과를 냈나요?", "어떤 역량이 돋보였나요?", "향후 어떤 목표를 세웠나요?" 같은 식입니다.

4. answers: 위 3개 질문에 대한 답변을 문서 내용에서 추출해 각각 3~5문장으로 작성하세요.
   - 반드시 문서에 있는 내용만 사용하세요.

5. examples: 이 문서와 같은 분야/주제에서 나올 수 있는 유사 사례 예시 3개를 생성하세요.
   - 이것만 창작이 허용됩니다.
   - 각 예시는 title, q1, q2, q3 필드를 포함합니다.
   - q1, q2, q3는 위에서 만든 questions의 1번, 2번, 3번 질문에 대한 답변입니다.
   - 첨부 문서의 내용을 그대로 반복하지 말고, 같은 분야에서 다른 상황/대상/결과를 다루세요.
   - 각 q1, q2, q3는 3~5문장으로 구체적이고 생생하게 작성하세요.
   - 독자가 "이 정도로 써야 좋은 사례가 되는구나" 하고 감을 잡을 수 있을 만큼 디테일하게 쓰세요.
   - 단순 요약이 아니라 실제 블로그 사례 원문처럼 현장감 있게 작성하세요.

반드시 순수 JSON만 출력하세요:
{{
  "summary": "문서 요약",
  "suggested_title": "사례명 제안",
  "questions": ["질문1", "질문2", "질문3"],
  "answers": ["답변1", "답변2", "답변3"],
  "examples": [
    {{"title": "예시 제목1", "q1": "질문1에 대한 구체적 답변 (3~5문장)", "q2": "질문2에 대한 구체적 답변 (3~5문장)", "q3": "질문3에 대한 구체적 답변 (3~5문장)"}},
    {{"title": "예시 제목2", "q1": "...", "q2": "...", "q3": "..."}},
    {{"title": "예시 제목3", "q1": "...", "q2": "...", "q3": "..."}}
  ]
}}
"""


if __name__ == "__main__":
    if len(sys.argv) <= 1:
        print(json.dumps({"error": "No data provided."}, ensure_ascii=False))
        sys.exit(0)

    response = ""
    try:
        raw = base64.b64decode(sys.argv[1]).decode("utf-8")
        data = json.loads(raw)
        prompt_text = build_analyze_prompt(data)
        response = generate_response(prompt_text)
        cleaned = strip_markdown_codeblock(response)
        parsed = json.loads(cleaned)
        print(json.dumps({"response": parsed}, ensure_ascii=False))
    except json.JSONDecodeError:
        print(json.dumps({"error": "JSON 파싱 실패", "raw": response}, ensure_ascii=False))
    except Exception as e:
        print(json.dumps({"error": str(e)}, ensure_ascii=False))
